MIG-119 - Introduce migration pipeline

pranavgaikwad opened the pull request:
This PR is part of the progress reporting improvements in mig-controller. It introduces a new field in `Phase` struct to assign a `Phase` with a logically related high level _Step_:


```go
// Phase
type Phase struct {
	[...]
        Step string
}
```

High level steps introduced:

```
Prepare
Backup
StageBackup
StageRestore
Restore
Final
```

## Pipeline

This PR also introduces a new field in _MigMigrationStatus_ struct to capture the progress of an ongoing migration through different _Steps_.

## Phase descriptions

Phase description work is cherry-picked from this PR: https://github.com/konveyor/mig-controller/pull/656

Originally, the proposed field for description was `Status.Description`. This PR instead uses the `message` field of current step in the pipeline. 

### Pipeline in action

Example 1: Failed Migration

```yml
status:
  conditions:
  - category: Advisory
    durable: true
    lastTransitionTime: "2020-10-29T21:37:01Z"
    message: '[1] Stage pods created.'
    status: "True"
    type: StagePodsCreated
  - category: Advisory
    durable: true
    lastTransitionTime: "2020-10-29T21:37:32Z"
    message: 'The migration has failed.  See: Errors.'
    reason: StageBackup
    status: "True"
    type: Failed
  errors:
  - 'PVC: mssql-persistent/mssql-pvc, deleted.'
  itinerary: Failed
  observedDigest: 9834d071975562d5e2c2eb855bca6950711ded8a0e45af5307fa56cd0f5ba3c7
  phase: Completed
  pipeline:
  - completed: "2020-10-29T21:35:53Z"
    message: Completed
    name: Prepare
    started: "2020-10-29T21:33:02Z"
  - completed: "2020-10-29T21:36:58Z"
    message: Completed
    name: Backup
    started: "2020-10-29T21:35:53Z"
  - completed: "2020-10-29T21:37:32Z"
    failed: true
    message: Failed
    name: StageBackup
    started: "2020-10-29T21:36:58Z"
  - message: Not started
    name: StageRestore
  - message: Not started
    name: Restore
  - completed: "2020-10-29T21:37:34Z"
    message: Completed
    name: Final
    started: "2020-10-29T21:37:32Z"
  startTimestamp: "2020-10-29T21:33:02Z"
```

Example 2 : Successful migration

```yml
status:
  conditions:
  - category: Advisory
    durable: true
    lastTransitionTime: "2020-10-29T21:53:45Z"
    message: The migration has completed successfully.
    reason: Final
    status: "True"
    type: Succeeded
  itinerary: Final
  observedDigest: 9834d071975562d5e2c2eb855bca6950711ded8a0e45af5307fa56cd0f5ba3c7
  phase: Completed
  pipeline:
  - completed: "2020-10-29T21:48:14Z"
    message: Completed
    name: Prepare
    started: "2020-10-29T21:45:11Z"
  - completed: "2020-10-29T21:48:49Z"
    message: Completed
    name: Backup
    started: "2020-10-29T21:48:14Z"
  - completed: "2020-10-29T21:50:22Z"
    message: Completed
    name: StageBackup
    started: "2020-10-29T21:48:49Z"
  - completed: "2020-10-29T21:52:24Z"
    message: Completed
    name: StageRestore
    started: "2020-10-29T21:50:22Z"
  - completed: "2020-10-29T21:53:45Z"
    message: Completed
    name: Restore
    started: "2020-10-29T21:52:24Z"
  - completed: "2020-10-29T21:53:45Z"
    message: Completed
    name: Final
    started: "2020-10-29T21:53:45Z"
  startTimestamp: "2020-10-29T21:45:11Z"
```

## QE Test Results 

Commit [#ca28ee8](https://github.com/konveyor/mig-controller/pull/706/commits/ca28ee85c9fdca1a54fe56566291bffe25c93ac7)

Image: quay.io/pranavgaikwad/mig-controller:latest

```
EXECUTION RESULT: OK (smoke/ocp-29946-hooksfailure.yml)
EXECUTION RESULT: OK (smoke/ocp-25986-maxpods.yml)
EXECUTION RESULT: OK (smoke/ocp-29918-hooks.yml)
EXECUTION RESULT: OK (smoke/ocp-24871-datagrid.yml)
EXECUTION RESULT: FAILURE (smoke/ocp-24769-cakephp.yml)
EXECUTION RESULT: FAILURE (smoke/ocp-34867-imagestreamsstage.yml)
EXECUTION RESULT: OK (smoke/ocp-24659-mysql.yml)
EXECUTION RESULT: OK (smoke/ocp-30240-datavalidation.yml)
EXECUTION RESULT: OK (smoke/ocp-25021-cronjob.yml)
EXECUTION RESULT: OK (smoke/ocp-27335-velero-version.yml)
EXECUTION RESULT: OK (smoke/ocp-32077-empty-pvc.yml)
EXECUTION RESULT: OK (smoke/ocp-24995-role.yml)
EXECUTION RESULT: OK (smoke/ocp-31358-restartrestic-newer-than-310.yml)
EXECUTION RESULT: OK (smoke/ocp-26160-max-pvs.yml)
EXECUTION RESULT: OK (smoke/ocp-24686-project.yml)
EXECUTION RESULT: OK (smoke/ocp-25022-cronjob-quiesced.yml)
EXECUTION RESULT: OK (smoke/ocp-24706-basicvolmig.yml)
EXECUTION RESULT: OK (smoke/ocp-24797-mongodb.yml)
EXECUTION RESULT: OK (smoke/ocp-29073-internalis.yml)
EXECUTION RESULT: OK (smoke/ocp-34821-increase-data.yml)
EXECUTION RESULT: OK (smoke/ocp-29947-customhooks.yml)
EXECUTION RESULT: OK (smoke/ocp-24787-redis.yml)
EXECUTION RESULT: OK (smoke/ocp-28967-migplan-naming.yml)
EXECUTION RESULT: OK (smoke/ocp-24730-django.yml)
EXECUTION RESULT: OK (smoke/ocp-34824-decrease-data.yml)
EXECUTION RESULT: OK (smoke/ocp-31309-quotanoattach.yml)
EXECUTION RESULT: OK (smoke/ocp-34703-crd.yml)
EXECUTION RESULT: FAILURE (smoke/ocp-24997-configmap.yml)
EXECUTION RESULT: OK (smoke/ocp-33378-max-pvs-equalquota.yml)
EXECUTION RESULT: OK (smoke/ocp-26032-maxns.yml)
EXECUTION RESULT: OK (smoke/ocp-25090-jobs.yml)
EXECUTION RESULT: OK (smoke/ocp-32189-cronjob-with-pvc.yml)
EXECUTION RESULT: FAILURE (smoke/ocp-28021-istags.yml)
EXECUTION RESULT: OK (smoke/ocp-29871-resquota.yml)
EXECUTION RESULT: OK (smoke/ocp-33393-bz1845569.yml)
EXECUTION RESULT: OK (smoke/ocp-33423-hooksnosa.yml)
EXECUTION RESULT: OK (smoke/ocp-32834-pvc-terminating.yml)
EXECUTION RESULT: OK (smoke/ocp-25000-sets.yml)
EXECUTION RESULT: OK (smoke/ocp-31339-pvc-not-connected.yml)
EXECUTION RESULT: OK (smoke/ocp-25212-initcont.yml)
EXECUTION RESULT: OK (smoke/ocp-33375-maxpods-equalquota.yml)
EXECUTION RESULT: OK (smoke/ocp-33379-maxns-equalquota.yml)
EXECUTION RESULT: OK (smoke/ocp-24872-mssql.yml)
```

djwhatle commented on pkg/controller/migmigration/migrate.go:
```diff
@@ -92,15 +91,12 @@ func (r *ReconcileMigMigration) migrate(migration *migapi.MigMigration) (time.Du
+		Reason:   task.Step,
 		Category: Advisory,
-		Message:  message,
+		Message:  PhaseDescriptions[task.Phase],
```
We need to verify that a PhaseDescription for the Phase exists before setting it here.

djwhatle commented on pkg/controller/migmigration/task.go:
```diff
@@ -89,139 +99,201 @@ type Itinerary struct {
+				{Name: DeleteRestores},
+				{Name: EnsureStagePodsDeleted, all: HasStagePods},
+				{Name: EnsureAnnotationsDeleted, any: HasPVs | HasISs},
+				{Name: DeleteMigrated},
```
@pranavgaikwad we've recently trimmed a few steps from the `CancelItinerary` with our rollback on demand changes. See the diff, we should match the phase list from master.

pranavgaikwad replied:
Thank you for pointing that. I missed this in a rebase

djwhatle commented on pkg/controller/migmigration/task.go:
```diff
@@ -89,139 +99,201 @@ type Itinerary struct {
+				{Name: EnsureAnnotationsDeleted, any: HasPVs | HasISs},
+				{Name: DeleteMigrated},
+				{Name: EnsureMigratedDeleted},
+				{Name: UnQuiesceApplications, all: Quiesce},
```
Same here, steps were trimmed from failed.

pranavgaikwad commented:
QE Test Results:

```
EXECUTION RESULT: OK (smoke/ocp-29946-hooksfailure.yml)
EXECUTION RESULT: OK (smoke/ocp-25986-maxpods.yml)
EXECUTION RESULT: OK (smoke/ocp-29918-hooks.yml)
EXECUTION RESULT: OK (smoke/ocp-24871-datagrid.yml)
EXECUTION RESULT: OK (smoke/ocp-24769-cakephp.yml)
EXECUTION RESULT: FAILURE (smoke/ocp-34867-imagestreamsstage.yml)
EXECUTION RESULT: OK (smoke/ocp-24659-mysql.yml)
EXECUTION RESULT: FAILURE (smoke/ocp-30240-datavalidation.yml)
EXECUTION RESULT: FAILURE (smoke/ocp-25021-cronjob.yml)
EXECUTION RESULT: OK (smoke/ocp-27335-velero-version.yml)
EXECUTION RESULT: OK (smoke/ocp-32077-empty-pvc.yml)
EXECUTION RESULT: OK (smoke/ocp-24995-role.yml)
EXECUTION RESULT: OK (smoke/ocp-31358-restartrestic-newer-than-310.yml)
EXECUTION RESULT: OK (smoke/ocp-26160-max-pvs.yml)
EXECUTION RESULT: OK (smoke/ocp-24686-project.yml)
EXECUTION RESULT: OK (smoke/ocp-25022-cronjob-quiesced.yml)
EXECUTION RESULT: OK (smoke/ocp-24706-basicvolmig.yml)
EXECUTION RESULT: OK (smoke/ocp-24797-mongodb.yml)
EXECUTION RESULT: OK (smoke/ocp-29073-internalis.yml)
EXECUTION RESULT: FAILURE (smoke/ocp-34821-increase-data.yml)
EXECUTION RESULT: FAILURE (smoke/ocp-29947-customhooks.yml)
EXECUTION RESULT: FAILURE (smoke/ocp-24787-redis.yml)
EXECUTION RESULT: FAILURE (smoke/ocp-28967-migplan-naming.yml)
EXECUTION RESULT: OK (smoke/ocp-24730-django.yml)
EXECUTION RESULT: OK (smoke/ocp-34824-decrease-data.yml)
EXECUTION RESULT: OK (smoke/ocp-31309-quotanoattach.yml)
EXECUTION RESULT: OK (smoke/ocp-34703-crd.yml)
EXECUTION RESULT: FAILURE (smoke/ocp-24997-configmap.yml)
EXECUTION RESULT: FAILURE (smoke/ocp-33378-max-pvs-equalquota.yml)
EXECUTION RESULT: OK (smoke/ocp-26032-maxns.yml)
EXECUTION RESULT: OK (smoke/ocp-25090-jobs.yml)
EXECUTION RESULT: OK (smoke/ocp-32189-cronjob-with-pvc.yml)
EXECUTION RESULT: FAILURE (smoke/ocp-28021-istags.yml)
EXECUTION RESULT: OK (smoke/ocp-29871-resquota.yml)
EXECUTION RESULT: OK (smoke/ocp-33393-bz1845569.yml)
EXECUTION RESULT: OK (smoke/ocp-33423-hooksnosa.yml)
EXECUTION RESULT: OK (smoke/ocp-32834-pvc-terminating.yml)
EXECUTION RESULT: OK (smoke/ocp-25000-sets.yml)
EXECUTION RESULT: OK (smoke/ocp-31339-pvc-not-connected.yml)
EXECUTION RESULT: OK (smoke/ocp-25212-initcont.yml)
EXECUTION RESULT: FAILURE (smoke/ocp-33375-maxpods-equalquota.yml)
EXECUTION RESULT: FAILURE (smoke/ocp-33379-maxns-equalquota.yml)
EXECUTION RESULT: OK (smoke/ocp-24872-mssql.yml)
```

Some of the above failures were due to api server timeouts, others were just due to migration taking longer time than the time configured for retries. I ran failed tests again one by one, all but one succeded when tried one at a time. `ocp-28021-istags.yml` failed in both cases.

alaypatel07 commented:
> ocp-28021-istags.yml failed in both cases.

This test case is expected to fail. There is an open bug for this.

alaypatel07 commented on pkg/controller/migmigration/description.go:
```diff
@@ -0,0 +1,53 @@
+	"InitialBackupFailed":             "Migration failed during initial Velero backup.",
+	"AnnotateResources":               "Adding migration annotations and labels to PVs, PVCs, Pods, ImageStreams, and Namespaces. Annotations and labels provide migration instructions to Velero, Velero Plugins and Restic.",
+	"EnsureStagePodsFromRunning":      "Creating Stage Pods and mounting PVC data from running Pods.",
+	"EnsureStagePodsFromTemplates":    "Creating Stage Pods and mounting PVC data from DeploymentTemplates, DeploymentConfigs, ReplicationControllers, DaemonSets, StatefulSets, ReplicaSets, CronJobs and Jobs.",
```
Suggest rewording this to: "Creating Stage Pods and mounting PVC using pod templates from Deployments, DeploymentConfigs, ReplicationControllers, DaemonSets, StatefulSets, ReplicaSets, CronJobs and Jobs."

djwhatle replied:
👍  on the change, just need to capitalize `Pod` per UX/docs team guidance on these descriptions.

pranavgaikwad commented:
@djwhatle I added old `Phase: x/y` progress report back to the `message` field along with descriptions. The goal is to not break UI until new changes come in. I added a TODO comment and will create a new issue to remove those when UI changes take place.

djwhatle commented:
@pranavgaikwad I was going to recommend considering that to enable earlier merge, 👍

pranavgaikwad commented:
UI seems to be working fine with the current message field:

```
  conditions:
  - category: Advisory
    lastTransitionTime: "2020-10-26T21:00:46Z"
    message: 'Phase: 3/34, Starting refresh on MigPlan, MigStorage and MigCluster
      resources'
    reason: Prepare
    status: "True"
    type: Running
  - category: Required
    lastTransitionTime: "2020-10-26T21:00:31Z"
    message: The migration is ready.
    status: "True"
    type: Ready
```

![Screenshot from 2020-10-26 17-01-00](https://user-images.githubusercontent.com/9839757/97228252-04aa9000-17ad-11eb-81a0-d88ec59fb056.png)

alaypatel07 commented on pkg/apis/migration/v1alpha1/timed.go:
```diff
@@ -0,0 +1,53 @@
+	return r.Completed != nil
+}
+
+func (r *Timed) now() *meta.Time {
```
is this function really needed? Can we not use meta.Now() where ever this function is intended to use?

alaypatel07 replied:
since we need this as it, marking the comment resolved

alaypatel07 commented on pkg/apis/migration/v1alpha1/timed.go:
```diff
@@ -0,0 +1,53 @@
+}
+
+// MarkReset mark as reset.
+func (r *Timed) MarkReset() {
```
is this function really needed? will creating a new `Timed{}` object instead achieve the same result?

pranavgaikwad replied:
@alaypatel07 we need this timed.go file as-is for future pipeline improvements. This is taken from virt-controller

alaypatel07 commented on pkg/apis/migration/v1alpha1/timed.go:
```diff
@@ -0,0 +1,53 @@
+func (r *Timed) MarkStarted() {
+	if r.Started == nil {
+		r.Started = r.now()
+		r.Completed = nil
```
pull this out of the if condition? Completed should be reset `nil` every time this function is called right?

pranavgaikwad replied:
not really, MarkStarted() should only mark a Step started, it shouldn't worry about whether a step is completed or not. 
but when it does start a step, it should make sure the completed is not set by mistake

alaypatel07 replied:
aah I see. So you want to make MarkStarted idempotent, cool

djwhatle commented on pkg/controller/migmigration/task.go:
```diff
@@ -44,6 +44,7 @@ const (
 	EnsureStagePodsFromTemplates    = "EnsureStagePodsFromTemplates"
 	EnsureStagePodsFromOrphanedPVCs = "EnsureStagePodsFromOrphanedPVCs"
 	StagePodsCreated                = "StagePodsCreated"
+	StagePodsFailed                 = "StagePodsFailed"
```
I believe this was removed from master

djwhatle reviewed (approved):
LGTM! Pipeline view looks good.

alaypatel07 reviewed (commented):
Added some initial comments, will continue the review in some time. In general, I have a design question. I could envision an alternative design of the same end goal which would not involve a lot of the Itinerary switch phase switching logic:

1. Introduce a new CR, MigMigrationPipeline. This CR is intended to be used by the UI only
2. The controller for this new CR will watch on MigMigration and MigMigrationPipeline objects.
3. Any time MigMigration CR changes its phase from in its Itinerary, this new controller will have a mapping of Itinerary phase to Step. The MigMigrationPipeline CR will update its status to reflect the right step for the phase.
4. The UI will only query the MigMigrationPipeline for pipeline view and MigMigration CR for a detailed view.

Questions:
1. Was this considered and rejected?
2. Do you say any major problem with the alternative?

I will say this, I haven't spent as much time thinking about this as Pranav has so I might be missing a lot of implementation detail. Also, know that this is blocking UI for 1.4.0. But just some thoughts to explore

djwhatle commented:
@alaypatel07 I don't like the idea of creating a degree of separation between the migmigration object and the user being able to see progress.

alaypatel07 commented:
@djwhatle that is why we have the UI for it :) if you are really at the CLI for debugging, you would really be looking at the phase level detail rather than the pipeline view right?

djwhatle commented:
> you would really be looking at the phase level detail and ignoring the MigMigrationPipeline right? The step is just a syntax wrapping to make the learning experience smother in UI

No, this is not true. The pipeline view is where progress on Backups, Restores, PVBs, PVRs will be reported.

alaypatel07 commented:
> No, this is not true. The pipeline view is where progress on Backups, Restores, PVBs, PVRs will be reported.

In the alternate design that could still be reported at phase level granularity and carried forward to the pipeline for the UI. So a user will still be able to see the progress of the phase that includes velero resource. Anyway, too late for the design discussion I guess, I would be better off focussing on making this one ready for merge

alaypatel07 reviewed (approved):
VISIACK.

Didnt get a chance to try this but assuming others have.

djwhatle reviewed (approved):
LGTM. Just tested this. It works well and doesn't break mig-ui while we transition over to reading data out of pipeline. Nice work integrating this @pranavgaikwad
